File: VisiumHD/03_UCD.R
# Visium HD 16 µm — Tumor-Only UCD Input Preparation
# Exports counts + coordinates for tumor bins only (tumor-of-origin input).
# Requires annotated objects from 02_tumor_annotation.R. Then run 03_UCD.py.

library(Seurat)
library(Matrix)

# Parameters
assay_name <- "Spatial.016um"
image_name <- "slice1.016um"

# Sample paths
samples <- list(

  CUP_062 = list(
    sample_id      = "L02",
    annotated_rds  = "./0.visiumHD/5.UCD/UCD_annotated_objects_16um/L02_16um_seu_ucd_annotated.rds",
    ucd_input_dir  = "./0.visiumHD/5.UCD/UCD_input_L02_16um_TumorOnly"
  ),

  CUP_138 = list(
    sample_id      = "L03",
    annotated_rds  = "./0.visiumHD/5.UCD/UCD_annotated_objects_16um/L03_16um_seu_ucd_annotated.rds",
    ucd_input_dir  = "./0.visiumHD/5.UCD/UCD_input_L03_16um_TumorOnly"
  ),

  CUP_204_1O = list(
    sample_id      = "L04_1O",
    annotated_rds  = "./0.visiumHD/5.UCD/UCD_annotated_objects_16um/L04_1O_16um_seu_ucd_annotated.rds",
    ucd_input_dir  = "./0.visiumHD/5.UCD/UCD_input_L04_1O_16um_TumorOnly"
  )
)

# Export tumor-only UCD inputs
for (sample_name in names(samples)) {

  s <- samples[[sample_name]]
  message("\n=== ", sample_name, " (", s$sample_id, ") ===")

  # Load annotated Seurat object
  seu <- readRDS(s$annotated_rds)
  DefaultAssay(seu) <- assay_name

  # Subset to Tumor bins only
  tumor_bins <- colnames(seu)[
    !is.na(seu$ucd_compartment) & seu$ucd_compartment == "Tumor"]

  if (length(tumor_bins) == 0) {
    warning("No Tumor bins found for ", sample_name, " — skipping.")
    next
  }

  seu_tumor <- subset(seu, cells = tumor_bins)

  # Counts matrix (genes x bins)
  counts <- GetAssayData(seu_tumor, assay = assay_name, layer = "counts")

  # Spatial coordinates
  coords <- GetTissueCoordinates(seu_tumor[[image_name]])
  coords <- coords[colnames(seu_tumor), , drop = FALSE]
  coord_export <- data.frame(x = coords$x, y = coords$y, row.names = rownames(coords))

  # Write outputs
  dir.create(s$ucd_input_dir, showWarnings = FALSE, recursive = TRUE)
  write.csv(as.matrix(counts),  file.path(s$ucd_input_dir, "counts.csv"))
  write.csv(coord_export,        file.path(s$ucd_input_dir, "coord.csv"))

  message(sprintf(
    "  Tumor bins: %d / %d total (%.1f%%)",
    length(tumor_bins), ncol(seu), 100 * length(tumor_bins) / ncol(seu)))
  message(sprintf(
    "  Exported: %d genes x %d bins", nrow(counts), ncol(counts)))
  message("  Written to: ", s$ucd_input_dir)
}

message("\nTumor-only UCD inputs ready. Run 03_UCD.py to execute UCD.")

File: VisiumHD/03_UCD.py
#!/usr/bin/env python3
# ============================================================
# Visium HD 16 µm — UniCell Deconvolve (UCD) Run
# ============================================================
# Runs UCD deconvolution on tumor-only spatial bins for all
# three CUP samples. Produces raw, primary, and cancer-level
# deconvolution scores used for tumor-origin prediction.
#
# Prerequisite: tumor-only counts.csv and coord.csv must exist
# in each sample's input directory. These are produced by
# 03_UCD.R.
#
# Authentication: set the UCD_TOKEN environment variable
# before running:
#   export UCD_TOKEN="your_token_here"
#   python 03_UCD.py
#
# Input:  UCD_input_*_TumorOnly/counts.csv + coord.csv
# Output: UCD_output_*_TumorOnly/ucdbase_{raw,primary,cancer}.csv
#         UCD_output_*_TumorOnly/adata_vis.h5ad
# ============================================================

import os
import sys
from pathlib import Path

import pandas as pd
from anndata import AnnData
import ucdeconvolve as ucd

# ── Authentication ────────────────────────────────────────────
token = os.environ.get("UCD_TOKEN")
if not token:
    sys.exit(
        "Error: UCD_TOKEN environment variable is not set.\n"
        "Set it with: export UCD_TOKEN='your_token_here'"
    )

# ── Paths ─────────────────────────────────────────────────────
base_dir = Path(__file__).resolve().parent

samples = {
    "CUP_062": {
        "sample_id":  "L02_16um",
        "input_dir":  base_dir / "UCD_input_L02_16um_TumorOnly",
        "output_dir": base_dir / "UCD_output_L02_16um_TumorOnly",
    },
    "CUP_138": {
        "sample_id":  "L03_16um",
        "input_dir":  base_dir / "UCD_input_L03_16um_TumorOnly",
        "output_dir": base_dir / "UCD_output_L03_16um_TumorOnly",
    },
    "CUP_204_1O": {
        "sample_id":  "L04_1O_16um",
        "input_dir":  base_dir / "UCD_input_L04_1O_16um_TumorOnly",
        "output_dir": base_dir / "UCD_output_L04_1O_16um_TumorOnly",
    },
}

# ── Authenticate once ─────────────────────────────────────────
ucd.api.authenticate(token)

# ── Process each sample ───────────────────────────────────────
for sample_name, s in samples.items():

    print(f"\n=== {sample_name} ({s['sample_id']}) ===")

    counts_path = s["input_dir"] / "counts.csv"
    coord_path  = s["input_dir"] / "coord.csv"

    if not counts_path.exists():
        raise FileNotFoundError(f"Counts file not found: {counts_path}")
    if not coord_path.exists():
        raise FileNotFoundError(f"Coordinate file not found: {coord_path}")

    # Load input files
    counts = pd.read_csv(counts_path, index_col=0)
    coord  = pd.read_csv(coord_path,  index_col=0)

    # Align barcodes
    common_bins = counts.columns.intersection(coord.index)
    if len(common_bins) == 0:
        raise ValueError(
            f"No matching barcodes between counts and coordinates for {sample_name}."
        )
    counts = counts.loc[:, common_bins]
    coord  = coord.loc[common_bins, :]

    print(f"  Counts : {counts.shape[0]} genes x {counts.shape[1]} bins")
    print(f"  Matched: {len(common_bins)} bins")

    # Build AnnData (bins x genes)
    adata = AnnData(counts.T)
    adata.var["SYMBOL"]      = adata.var_names
    adata.obs["array_row"]   = coord.loc[adata.obs_names, "y"].values
    adata.obs["array_col"]   = coord.loc[adata.obs_names, "x"].values
    adata.obsm["spatial"]    = coord.loc[adata.obs_names, ["x", "y"]].to_numpy()
    adata.obs["sample"]      = s["sample_id"]

    s["output_dir"].mkdir(parents=True, exist_ok=True)

    # Run UCD deconvolution
    try:
        ucd.tl.base(adata)

        for category in ["raw", "primary", "cancer"]:
            result = ucd.utils.read_results(adata, category=category)
            out_path = s["output_dir"] / f"ucdbase_{category}.csv"
            result.to_csv(out_path, sep="\t")
            print(f"  Saved {category}: {out_path}")

        adata.write(s["output_dir"] / "adata_vis.h5ad")
        print(f"  Saved AnnData: {s['output_dir'] / 'adata_vis.h5ad'}")
        print(f"  UCD complete for {sample_name}.")

    except Exception as e:
        print(f"\n  UCD failed for {sample_name}.")
        print(f"  {type(e).__name__}: {e}")
        try:
            adata.write(s["output_dir"] / "adata_vis_partial.h5ad")
            print("  Partial AnnData saved.")
        except Exception:
            pass

print("\nUCD complete for all samples.")
